File: go/workflows/account_transfer_workflow.go
package workflows

import (
	"money-transfer-worker/activities"
	"money-transfer-worker/app"
	"money-transfer-worker/messages"
	"time"

	"github.com/google/uuid"
	"go.temporal.io/sdk/temporal"
	"go.temporal.io/sdk/workflow"
)

func AccountTransferWorkflow(ctx workflow.Context, input app.TransferInput) (output *app.TransferOutput, err error) {
	name := workflow.GetInfo(ctx).WorkflowType.Name
	logger := workflow.GetLogger(ctx)
	logger.Info("Account Transfer workflow started", "name", name)

	activityOptions := workflow.ActivityOptions{
		StartToCloseTimeout: 5 * time.Second,
		RetryPolicy: &temporal.RetryPolicy{
			InitialInterval:    1 * time.Second,
			BackoffCoefficient: 2.0,
			MaximumInterval:    30 * time.Second,
		},
	}
	ctx = workflow.WithActivityOptions(ctx, activityOptions)

	// Expose transfer status as a query
	ts, err := messages.SetQueryHandlerForStatus(ctx)
	if err != nil {
		return nil, err
	}

	var idempotencyKey string
	_ = workflow.SideEffect(ctx, func(ctx workflow.Context) interface{} {
		return uuid.New().String()
	}).Get(&idempotencyKey)

	// Validate
	err = workflow.ExecuteActivity(ctx, activities.Validate, input).Get(ctx, nil)
	if err != nil {
		return nil, err
	}
	updateProgress(ctx, 1, ts, 25)

	// Withdraw
	err = workflow.ExecuteActivity(ctx, activities.Withdraw, idempotencyKey, input.Amount, name).Get(ctx, nil)
	if err != nil {
		return nil, err
	}
	updateProgress(ctx, 3, ts, 50)

	// Deposit
	depositResponse := app.DepositResponse{}
	err = workflow.ExecuteActivity(ctx, activities.Deposit, idempotencyKey, input.Amount, name).Get(ctx, &depositResponse)
	if err != nil {
		return nil, err
	}
	updateProgress(ctx, 1, ts, 75)

	// Send Notification
	err = workflow.ExecuteActivity(ctx, activities.SendNotification, input).Get(ctx, nil)
	if err != nil {
		return nil, err
	}
	updateProgress(ctx, 1, ts, 100)

	output = &app.TransferOutput{
		DepositResponse: depositResponse,
	}
	return output, nil
}

func updateProgress(ctx workflow.Context, sleep int, ts *app.TransferStatus, progress int) {
	if sleep > 0 {
		duration := time.Duration(sleep) * time.Second
		workflow.Sleep(ctx, duration)
	}
	state := "running"
	if progress == 100 {
		state = "finished"
	}
	ts.TransferState = state
	ts.ProgressPercentage = progress
}

File: ruby/workflows/account_transfer_workflow.rb
# frozen_string_literal: true

require 'temporalio/workflow'
require_relative '../models/transfer_input'
require_relative '../models/transfer_output'
require_relative '../models/transfer_status'
require_relative '../models/deposit_response'

module Workflows
  class AccountTransferWorkflow < Temporalio::Workflow::Definition
    attr_reader :progress, :transfer_state, :deposit_response

    def initialize
      @progress = 0
      @transfer_state = ''
      @deposit_response = Models::DepositResponse.new(charge_id: '')
    end

    def execute(input)
      workflow_type = Temporalio::Workflow.info.workflow_type
      logger.info("Simple workflow started: #{workflow_type}, input: #{input.to_h}")

      idempotency_key = Temporalio::Workflow.random.uuid

      validate_transfer(input)
      update_progress(25, 1)

      withdraw_funds(idempotency_key, input.fetch('amount'), workflow_type)
      update_progress(50, 3)

      @deposit_response = deposit_funds(idempotency_key, input.fetch('amount'), workflow_type)
      update_progress(75, 1)

      send_notification(input)
      update_progress(100, 1, 'finished')

      Models::TransferOutput.new(deposit_response: @deposit_response).to_h
    end

    workflow_query(name: 'transferStatus')
    def query_transfer_status
      logger.info('Workflow has been queried')
      Models::TransferStatus.new(
        progress_percentage: progress,
        transfer_state: transfer_state,
        workflow_status: '',
        charge_result: deposit_response,
        approval_time: 0
      ).to_h
    end

    private

    def validate_transfer(input)
      Temporalio::Workflow.execute_activity(
        Activities::ValidateActivity,
        input,
        start_to_close_timeout: 5,
        retry_policy: Activities::BaseActivity::RETRY_POLICY
      )
    end

    def withdraw_funds(idempotency_key, amount, workflow_type)
      Temporalio::Workflow.execute_activity(
        Activities::WithdrawActivity,
        idempotency_key, amount,
        start_to_close_timeout: 5,
        retry_policy: Activities::BaseActivity::RETRY_POLICY
      )
    end

    def deposit_funds(idempotency_key, amount, workflow_type)
      Temporalio::Workflow.execute_activity(
        Activities::DepositActivity,
        idempotency_key, amount,
        start_to_close_timeout: 5,
        retry_policy: Activities::BaseActivity::RETRY_POLICY
      )
    end

    def send_notification(input)
      Temporalio::Workflow.execute_activity(
        Activities::SendNotificationActivity,
        input,
        start_to_close_timeout: 5,
        retry_policy: Activities::BaseActivity::RETRY_POLICY
      )
    end

    def update_progress(progress, sleep_duration, state = 'running')
      sleep(sleep_duration) if sleep_duration > 0
      @transfer_state = state
      @progress = progress
    end

    def logger
      @logger ||= Temporalio::Workflow.logger
    end
  end
end
